### ISSUE TITLE
How install this tool?

### ISSUE THREAD

**Issue by uschi1** [vincentbitter/koreader-calibre-web-sync#5]
Hello, unfortunately I can't speak English (i am German) very well and am using Google to translate the text, so sorry for the bad English! I have set up a Docker on my Synology and installed my Calibre-web on it, how do I install this image so that I can sync the reading progress with Koreader? Is there a guide for this, I would be happy to hear it....

Ursula

**Comment by vincentbitter** [comment]
Hi Ursula, thank you for your interest in the tool. I am afraid I cannot give a better installation guidance than the one at https://github.com/vincentbitter/koreader-calibre-web-sync?tab=readme-ov-file#how-to-use, but I will try a step-by-step tutorial.

*Example installation details:*
NAS:
- IP: 10.0.0.0.100
Calibre-web:
- URL: http://10.0.0.100:3000
- Username: vincent
- Password: secret
Koreader-calibre-web-sync:
- URL: http://10.0.0.100:5000

If the Docker image is running:
- Open KOReader
- Go to the Tools menu on top
- Click "Progress sync"
- Click "Custom sync server"
- Fill in the URL: http://10.0.0.100:5000
- Save and go back to the "Progress sync" menu
- Click "Register / Login"
- Fill in the Username: vincent:secret@http://10.0.0.100:3000
- Fill in the Password: random (this value will be ignored, so fill in anything you like)
- Save

Now when you finished a book, it will send the 100% value to the koreader-calibre-web-sync server which should forward it to Calibre-web and mark the book as read.

Hope this works for you!

**Comment by uschi1** [comment]
Kann ich deutsch mit dir schreiben?

**Comment by uschi1** [comment]
Ich versuche es auf deutsch.....

Ich habe bereits einen Docker auf meiner Synology laufen, hier habe ich das Image "Calibre-web" installiert und das läuft auch Super!
Was ich jetzt nicht verstehe, muss ich jetzt im gleichen Docker dein Image installieren?

**Comment by uschi1** [comment]
Ich habe mal zwei Screenshots gemacht von der Installation von deinem Image bei mir auf dem Docker von Calibre-web....
![Screenshot_20241024_153558_Opera](https://github.com/user-attachments/assets/36fe2405-b135-49bc-bbc6-078667ebc12e)
![Screenshot_20241024_153524_Opera](https://github.com/user-attachments/assets/20a98030-063f-4de6-8254-ad48becde695)

**Comment by vincentbitter** [comment]
I understand German, but will reply in English, as Google Translate is better in German than I am 😄 

You must run both Docker files (koreader-calibre-web-sync and calibre-web) in parallel. So, both images should be running on your NAS.

In your last screenshot I see you are messing up with the port numbers. If possible, I recommend creating a portmapping (Port-Einstellungen) from TCP port 5000 to port 5000 instead of 84 to 8084. Also, in the environment variables (Umgebung) you have to set `ASPNETCORE_HTTP_PORTS` to 5000 instead of 8080. And leave `ASPNETCORE_URLS` set to `http://*:5000`.

**Comment by vincentbitter** [comment]
Ah, my bad! I see you are running DSM on port 5000, so the above won't work. Let me rephrase:

In your last screenshot I see you are messing up with the port numbers. I recommend creating a portmapping (Port-Einstellungen) from TCP port 84 to port 5000 instead of 84 to 8084. Also, in the environment variables (Umgebung) you have to set ASPNETCORE_HTTP_PORTS to 5000 instead of 8080. And leave ASPNETCORE_URLS set to http://*:5000.

The URL for the custom sync server should then be like http://mynasip:84

**Comment by uschi1** [comment]
Okay wenn ich das ändere kommt diese Fehlermeldung...
![Screenshot_20241024_160739_Opera](https://github.com/user-attachments/assets/5e047ec7-c317-424d-b13e-b6b9b3419f80)
![Screenshot_20241024_160752_Opera](https://github.com/user-attachments/assets/82951899-9eb8-4e71-a05e-1671f2f3bd74)

**Comment by uschi1** [comment]
![Screenshot_20241024_160843_Opera](https://github.com/user-attachments/assets/ad9c39cc-8faa-4601-ad17-95e222b0fe67)

**Comment by vincentbitter** [comment]
The port mapping (Port-Einstellungen) must be exactly the other way around. You are now trying to map public port 5000 to private port 84, while you should map public port 84 to private port 5000.

**Comment by uschi1** [comment]
Oje, jetzt verstehe ich gar nichts mehr.....
Hier mal noch ein Screenshot von meiner Calibre-web installation

![Uploading Screenshot_20241024_162139_Opera.jpg…]()

**Comment by vincentbitter** [comment]
Sorry, your image is not working

**Comment by uschi1** [comment]
geht es jetzt?

![calibre-web installation](https://github.com/user-attachments/assets/646cc2b1-5831-4fff-9491-15f9acc025a3)

**Comment by uschi1** [comment]
Noch etwas.... mein Calibre-web ist bei mir nur intern unter http://192.168.178.34:83 erreichbar

**Comment by vincentbitter** [comment]
Given that is the IP address of your NAS, the instructions should be:

Example installation details:
NAS:

IP: 192.168.178.34
Calibre-web:
URL: http://192.168.178.34:83/
Username: vincent
Password: secret
Koreader-calibre-web-sync:
URL: http://192.168.178.34:84/

Open KOReader
Go to the Tools menu on top
Click "Progress sync"
Click "Custom sync server"
Fill in the URL: http://192.168.178.34:84/
Save and go back to the "Progress sync" menu
Click "Register / Login"
Fill in the Username: vincent:secret@http://192.168.178.34:83/ (please replace 'vincent' and 'secret' with your own user/pass of Calibre-web)
Fill in the Password: random (this value will be ignored, so fill in anything you like)
Save

**Comment by uschi1** [comment]
Ist die Portfreigabe so richtig? Kann ich das irgendwie testen, ob es funktioniert?

![Screenshot_20241025_154140_Opera](https://github.com/user-attachments/assets/c16cdf6f-2ba1-44ea-8577-159bb981f3f9)

**Comment by vincentbitter** [comment]
Yes! You can validate by going to http://192.168.178.34:84/swagger/index.html which should give a white page with "Swagger UI" as the title of your browser tab.

**Comment by uschi1** [comment]
Leider klappt das nicht, also ist doch noch etwas falsch....😪


![Screenshot_20241025_181934_Opera](https://github.com/user-attachments/assets/6a72a63d-2243-4ec6-9cbd-57fbca23c3e1)

**Comment by uschi1** [comment]
Meinst du so?


![Screenshot_20241025_182603_Opera](https://github.com/user-attachments/assets/67c83dbb-9fdb-469b-a139-
945263d2ae50)

**Comment by uschi1** [comment]
![Screenshot_20241025_182603_Opera](https://github.com/user-attachments/assets/5dea24d2-b77e-452d-9953-3f1f2fbf5e6f)

**Comment by uschi1** [comment]
Ist das so richtig?

**Comment by vincentbitter** [comment]
Yes! The Docker container is reachable, so now configure your Koreader as described and you're good to go!

**Comment by uschi1** [comment]
Super, Danke schon mal für die Hilfe..... Den Rest mache ich dann am Wochenende, ich bin gespannt, ob es funktioniert!

**Comment by uschi1** [comment]
Hallo nochmal, ich habe jetzt alles eingetragen, aber noch eine Frage....
Wenn ich mehrere Geräte mit Koreader verwende (Handy/Tablet) und diesen Server eintrage, wird da auch der Lesefortschritt zwischen den Geräten synhronisiert?

update: Habe es gerade am Handy probiert, aber da wird nicht synchronisiert .... Schade normalerweise muss man bei koreader Server die Geräte erst registrieren und dann werden die Geräte untereinander synchronisiert, aber wenn ich deinen Server eingebe und auf "registrieren" klicke kommt der Fehler "unbekannter Server-Fehler"...

**Comment by vincentbitter** [comment]
> Wenn ich mehrere Geräte mit Koreader verwende (Handy/Tablet) und diesen Server eintrage, wird da auch der Lesefortschritt zwischen den Geräten synhronisiert?

No, as Calibre-web does not have a field available to fill with the progress. And koreader-calibre-web-sync does not have storage to store the progress itself. Pull requests are welcome :)

> update: Habe es gerade am Handy probiert, aber da wird nicht synchronisiert .... Schade normalerweise muss man bei koreader Server die Geräte erst registrieren und dann werden die Geräte untereinander synchronisiert, aber wenn ich deinen Server eingebe und auf "registrieren" klicke kommt der Fehler "unbekannter Server-Fehler"...

You don't have to register a new account, as you already have a Calibre-web username/password. Does login also give a server error?

**Comment by uschi1** [comment]
> Sie müssen kein neues Konto registrieren, da Sie bereits einen Calibre-web-Benutzernamen/ein neues Passwort haben. Gibt die Anmeldung auch einen Serverfehler?

Nein, ich habe es noch nicht probiert,  ob dann tatsächlich das Buch als gelesen markiert wird, aber wenn ich die Lesefortschritte nicht mit anderen Geräten synchronisieren kann, muss ich überlegen, ob das für mich eine Gute Lösung ist!
Leider kann man beim Koreader ja nur einen Server zum synchen angeben 😥

**Comment by BLACK4585** [comment]
> No, as Calibre-web does not have a field available to fill with the progress. And koreader-calibre-web-sync does not have storage to store the progress itself. Pull requests are welcome :)

Hey @vincentbitter , just asking in between here. So the tool just syncs from Koreader to Calibre-Web the status of the book, so unread/read, but not the exact progress? But not vice versa?
And this because Calibre-Web nor koreader-calibre-web-sync have a storage for this?

So what needed to be implemented is a storage system for the progress?

**Comment by vincentbitter** [comment]
> > No, as Calibre-web does not have a field available to fill with the progress. And koreader-calibre-web-sync does not have storage to store the progress itself. Pull requests are welcome :)
> 
> Hey [@vincentbitter](https://github.com/vincentbitter) , just asking in between here. So the tool just syncs from Koreader to Calibre-Web the status of the book, so unread/read, but not the exact progress? But not vice versa? And this because Calibre-Web nor koreader-calibre-web-sync have a storage for this?
> 
> So what needed to be implemented is a storage system for the progress?

Exactly! Syncing Calibre-web status (read/unread) to Koreader would be possible without storing anything I guess, although the book status will only be received when the book is opened in Koreader, so to see if a book was already read, you first have to open the book in Koreader to conclude it skips to the last page. 

To store the progress (%), there are two options:

## Option 1
1. Add a data storage to koreader-calibre-web-sync that keeps track of the progress per book. This could be just a mounted folder in Docker with a json or SQLite file.
1. Create a repository to access the progress in https://github.com/vincentbitter/koreader-calibre-web-sync/tree/main/Infrastructure/Repositories
1. Change https://github.com/vincentbitter/koreader-calibre-web-sync/blob/main/Application/Syncs/UpdateProgress/UpdateProgressHandler.cs so it stores the progress on the data storage.
1. Implement https://github.com/vincentbitter/koreader-calibre-web-sync/blob/main/Application/Syncs/GetProgress/GetProgressHandler.cs

## Option 2
I just saw it is possible to add custom columns in your Calibre Database via Calibre, which do appear in the web interface of Calibre-Web. 
1. Make an instruction on how to create such custom column in Calibre for "Progress".
1. Allow specifying a target field in the config of koreader-calibre-web-sync (namely "Progress") to store the progress percentage.
1. Add progress to https://github.com/vincentbitter/koreader-calibre-web-sync/blob/main/Domain/Book.cs
1. Change https://github.com/vincentbitter/koreader-calibre-web-sync/blob/main/Application/Syncs/UpdateProgress/UpdateProgressHandler.cs so it sets the progress in the Book object and mark it as dirty to have it synced later on.
1. Add a method to update the progress in calibre-web in https://github.com/vincentbitter/koreader-calibre-web-sync/blob/main/Infrastructure/Services/CalibreWebService.cs similar to `MarkBookAsReadSync`
1. Make sure the new method is called in https://github.com/vincentbitter/koreader-calibre-web-sync/blob/main/Application/LibrarySync/CalibreWebSync.cs#L48
1. Make a way to retrieve the progress % from calibre-web during the sync in `UpdateBooksFromRemote`: https://github.com/vincentbitter/koreader-calibre-web-sync/blob/main/Application/LibrarySync/CalibreWebSync.cs#L59
1. Implement https://github.com/vincentbitter/koreader-calibre-web-sync/blob/main/Application/Syncs/GetProgress/GetProgressHandler.cs

Tricky part here will be that if you open the book, Koreader will immediately try to grab the progress % from koreader-calibre-web-sync, which doesn't have executed its background sync job by then. Issue with koreader-calibre-web-sync is that it currently does not have any persistence, so after a reboot, it lost the Calibre-web URL and credentials. These are only sent when the first API call is done from Koreader.

Now, that was a pretty long answer, but hopefully it motivates to start working on this 😄

**Comment by BLACK4585** [comment]
But a pretty good answer as a starting point! I personally would go with option one for two reasons. It is shorter so should be a lot easier to implement. And you don't need to fiddle around with the Calibre DB, it would just require implementing some type of a man in the middle storage. Being more a docker, decentralized guy I'd prefer an extra database rather then a folder mount/volume or file. You could then easily spin up a DB container. Also you can then throw everything in a docker-compose.yml and deploy the whole, much requested progess-sync for Calibre-Web with a click of a button.
Reading your answer, you don't sound like being motivated to implement this yourself? I would be interested in doing so in the future. The only problem is, that I never worked with C# before. Just with C, C++ and Java. But I think they meet somewhere in the middle, tending to Java.
I think I will look into this in (the near) future, since I'm currently working on another project. Besides that, moving the discussion about this topic in a different issue would be beneficial.

**Comment by uschi1** [comment]
> Aber eine recht gute Antwort als Ausgangspunkt! Ich persönlich würde mich aus zwei Gründen für Option eins entscheiden. Es ist kürzer und sollte daher viel einfacher umzusetzen sein. Und Sie müssen nicht an der Calibre DB herumfummeln, sondern müssten lediglich eine Art Mann im mittleren Speicher implementieren. Da ich eher ein Docker und ein dezentraler Typ bin, würde ich eine zusätzliche Datenbank bevorzugen und nicht eine Ordnermontage/-volume oder -datei. Man könnte dann ganz einfach einen DB-Container hochdrehen. Außerdem können Sie dann alles in ein Docker-Compose.yml werfen und die gesamte, sehr gefragte Progess-Synchronisierung für Calibre-Web per Knopfdruck bereitstellen. Wenn Sie Ihre Antwort lesen, hören Sie sich nicht danach an, motiviert zu sein, dies selbst umzusetzen? Ich wäre daran interessiert, dies in Zukunft zu tun. Das Problem ist nur, dass ich noch nie mit C# gearbeitet habe. Nur mit C, C++ und Java. Aber ich denke, sie treffen sich irgendwo in der Mitte und neigen zu Java. Ich denke, ich werde mich in (nächster) Zukunft damit befassen, da ich gerade an einem anderen Projekt arbeite. Darüber hinaus wäre es von Vorteil, die Diskussion zu diesem Thema in ein anderes Thema zu verlagern.

Das wäre Super, wenn es so etwas geben würde, ich habe das wieder deinstalliert, weil ich auch den Lesefirtschrutt mit anderen Geräten synchronisieren möchte und das geht hiermit leider nicht! Ich hatte mir dann auf meinem Docker einen Sync Server für koreader installiert, dass klappt hervorragend,  aber leider wird das Buch dann in Calibre-web nicht als "gelesen" markiert... 😞, man bräuchte beides "two in one:

Ich habe das Calibre-web auch auf dem Docker laufen!

**Comment by viniciuscmartins** [comment]
Is there a way to install it with Docker Compose? I use truenas scale and dockge

**Comment by vincentbitter** [comment]
> Is there a way to install it with Docker Compose? I use truenas scale and dockge

Yes, it is just one container, so no real benefit of using docker compose though. Untested:
```yaml
services:
  kcws:
    image: vincentbitter/koreader-calibre-web-sync:latest
    container_name: koreader
    ports:
      - "5000:5000"
    networks:
      - calibre
    restart: unless-stopped
networks:
  calibre:
    driver: bridge
```

Of you want to run it together with calibre-web, you might need to add that as a service as well (untested):
```yaml
services:
  calibre-web:
    image: lscr.io/linuxserver/calibre-web:latest
    container_name: calibre-web
    environment:
      - PUID=1000
      - PGID=1000
      - TZ=Etc/UTC
      - DOCKER_MODS=linuxserver/mods:universal-calibre #optional
      - OAUTHLIB_RELAX_TOKEN_SCOPE=1 #optional
    volumes:
      - /path/to/calibre-web/data:/config
      - /path/to/calibre/library:/books
    ports:
      - 8083:8083
    networks:
      - calibre
    restart: unless-stopped
  kcws:
    image: vincentbitter/koreader-calibre-web-sync:latest
    container_name: koreader
    ports:
      - "5000:5000"
    depends_on:
      - calibre-web
    networks:
      - calibre
    restart: unless-stopped
networks:
  calibre:
    driver: bridge
```

**Comment by viniciuscmartins** [comment]
Thanks! I'll try to run together with the calibre web and let you know.

**Comment by viniciuscmartins** [comment]
I was able to run with docker compose, I can access the swagger link.
At Koreader I was able to login at progress sync and it says it has pushed the data, but Calibre-web never updates. 
Is there a Log I can look into to see where is it going wrong?

**Comment by vincentbitter** [comment]
The tool does some logging to the console, so should be visible in your Docker container logs. Usually it has to do with the things mentioned under "Notes" in the "Setup" section on https://github.com/vincentbitter/koreader-calibre-web-sync/blob/main/README.md#setup
It could also be that the filename on your e-reader is different than the one generated by calibre-web. This is something I still need to improve, but it's difficult. I even thought about creating a plugin for Koreader itself, which would make things a lot easier.  But time...

**Comment by viniciuscmartins** [comment]
Calibre-Web has the following settings:
![Image](https://github.com/user-attachments/assets/6205d66f-7a18-485a-a33e-db186934628f)

Progress sync has Document Matching method as Filename. 

this is what I can see at dockge terminal:

> koreader                                                 | warn: Microsoft.AspNetCore.DataProtection.Repositories.FileSystemXmlRepository[60]
> koreader                                                 |       Storing keys in a directory '/app/.aspnet/DataProtection-Keys' that may not be persisted outside of the container. Protected data will be unavailable when container is destroyed. For more information go to https://aka.ms/aspnet/dataprotectionwarning
> koreader                                                 | warn: Microsoft.AspNetCore.DataProtection.KeyManagement.XmlKeyManager[35]
> koreader                                                 |       No XML encryptor configured. Key {3f3ebf77-ba70-xxxx-yyyy-2b7126f4886f} may be persisted to storage in unencrypted form.
> koreader                                                 | warn: Microsoft.AspNetCore.Hosting.Diagnostics[15]
> koreader                                                 |       Overriding HTTP_PORTS '8080' and HTTPS_PORTS ''. Binding to values defined by URLS instead 'http://*:5000'.
> koreader                                                 | info: Microsoft.Hosting.Lifetime[14]
> koreader                                                 |       Now listening on: http://[::]:5000
> koreader                                                 | info: Microsoft.Hosting.Lifetime[0]
> koreader                                                 |       Application started. Press Ctrl+C to shut down.
> koreader                                                 | info: Microsoft.Hosting.Lifetime[0]
> koreader                                                 |       Hosting environment: Production
> koreader                                                 | info: Microsoft.Hosting.Lifetime[0]
> koreader                                                 |       Content root path: /app
> koreader                                                 | info: System.Net.Http.HttpClient.Default.LogicalHandler[100]
> koreader                                                 |       Start processing HTTP request GET http://192.168.0.51:8083/login
> koreader                                                 | info: System.Net.Http.HttpClient.Default.ClientHandler[100]
> koreader                                                 |       Sending HTTP request GET http://192.168.0.51:8083/login
> koreader                                                 | info: System.Net.Http.HttpClient.Default.ClientHandler[101]
> koreader                                                 |       Received HTTP response headers after 68.1878ms - 200
> koreader                                                 | info: System.Net.Http.HttpClient.Default.LogicalHandler[101]
> koreader                                                 |       End processing HTTP request after 87.6985ms - 200
> koreader                                                 | info: System.Net.Http.HttpClient.Default.LogicalHandler[100]
> koreader                                                 |       Start processing HTTP request POST http://192.168.0.51:8083/login
> koreader                                                 | info: System.Net.Http.HttpClient.Default.ClientHandler[100]
> koreader                                                 |       Sending HTTP request POST http://192.168.0.51:8083/login
> koreader                                                 | info: System.Net.Http.HttpClient.Default.ClientHandler[101]
> koreader                                                 |       Received HTTP response headers after 405.8909ms - 200
> koreader                                                 | info: System.Net.Http.HttpClient.Default.LogicalHandler[101]
> koreader                                                 |       End processing HTTP request after 407.133ms - 200

When I click in KOreader "Push progress from this device now", I don`t see any logs at the terminal, not sure if I should. 

I`m also downloading the books from calibre-web using opds format, so that could be the case.

**Comment by vincentbitter** [comment]
It seems to login correctly, and will try to fetch book data if there is any progress. Are you sure you went to the last page of the book when pushing progress?
